### ISSUE TITLE
Package Uyuni for Debian

### ISSUE THREAD

**Issue by paususe** [openSUSE/mentoring#136]
**Project Title:** Package Uyuni for Debian

**Description:** Currently, Uyuni is only available and runs on openSUSE Leap. A port to CentOS is in the works. In order to make it more popular, let's package it for Debian (which should also run easily on Ubuntu). It's your choice to decide whether you want to use debbuild (i. e. build .deb packages from the same .spec files we use for openSUSE, that requires some black magic with OBS and it will not be accepted by Debian upstream) or create native Debian packaging.

**Deliverable:** A packaged, mostly-working Uyuni Server and its dependencies. It is not required to package or make it run 100% of the features during this GSoC project but at least the core functionality should work.

Required working features at the end of this project (definition of done):
* Content management: package sync, Content Lifecycle Management, etc (should work mostly out of the box, it's just a matter of packaging).
* Configuration management
* OpenSCAP scan
* Managing Salt clients

Nice to have features:
* Building OS images (no big problems expected)
* Building containers (no big problems expected)
* CVE scan (requires https://github.com/openSUSE/mentoring/issues/127)
* Managing traditional clients (it's mostly packaging but a lot of it)

**Mentor:** @paususe (inactive Debian Developer) and some other Debian Developer(s) currently (either currently employed by SUSE or interested in Uyuni, it'll be easy to find them)

**Skills:** Linux and packaging. Open Build Service knowledge a plus (but you can learn it on the way :-))

**Skill Level:** Easy to medium if you already know packaging. Please note Uyuni comprises a ton of packages and dependencies, so while this project is no superdifficult per se, it requires that you already know packaging and you a disciplined.

**Get started:** Tasks that mentors may want to suggest students so that they can start contributing to the code base (e.g. junior jobs, low hanging fruits, discussion on the mailing list)

- https://github.com/uyuni-project/uyuni/wiki
- https://wiki.debian.org/Packaging
- [Debian packaging tutorial](https://www.debian.org/doc/manuals/packaging-tutorial/packaging-tutorial.en.pdf)
- https://github.com/uyuni-project/uyuni/ (look at the .spec files, you will need to essentially convert that to Debian packaging)
- Uyuni development in no time: https://github.com/uyuni-project/uyuni/wiki/Uyuni-development-in-no-time
- Uyuni wiki: https://github.com/uyuni-project/uyuni/wiki

**Comment by thisisshub** [comment]
Hi @paususe,
I have a bit of experience on packaging with [build.opensuse.org](https://build.opensuse.org/project/show/home:thisisshub_:tests) and I believe I could work on this project. 

Doubts,
- The only deliverable is to make a `.deb` package of `uyuni`? 
- Other than the **Get Started** tasks is there anything I could do?

**Comment by drs-11** [comment]
Hi @paususe! I'm interested in this issue for GSoC. Currently I do not have any experience in packaging of any kind but am in the middle of reading the "[Guide for Debian Maintainers](https://www.debian.org/doc/manuals/debmake-doc/index.en.html)" and will try to write some small packages as practice. I know this alone won't be enough to prove I am capable of this issue so is there any sort of qualification task?

**Comment by paususe** [comment]
> Hi @paususe! I'm interested in this issue for GSoC. Currently I do not have any experience in packaging of any kind but am in the middle of reading the "[Guide for Debian Maintainers](https://www.debian.org/doc/manuals/debmake-doc/index.en.html)" and will try to write some small packages as practice. I know this alone won't be enough to prove I am capable of this issue so is there any sort of qualification task?

Hello @drs-11 !

The Guide for Debian Maintainers is good reading but it focuses a lot on the "Debian toolset for packaging". Since Uyuni is large (all in all, we are talking about 100 packges, give or take), my recommendation is to NOT use Debian native packaging (i. e. `debian` directory with `rules`, `control`, `copyright`, etc files) but to reuse the existing RPM packaging. You can do that by leveraging `debbuild`, which is available in OBS.
https://github.com/debbuild/debbuild

I would recommend you first read the Debian Policy Guide, which will tell you what are the fields a Debian package needs:
https://www.debian.org/doc/debian-policy/

Also, you need to get familiar with OBS, since it will be a must-have tool for you:

Create an OBS account, branch some RPM package (those which contain a .spec file), try to add new repositories (e. g. for new versions of the same Linux distribution, or for other RPM-based distributions), etc.

When you are comfortable with that, try to learn debbuild. What debbuild does is it creates Debian packages (.deb) from RPM package specifications (.spec). Word of warning: debbuild is underdocumented. The best way to learn about debbuild and how to use it is to look at packages which use it, e. g:
https://build.opensuse.org/package/show/home:pagarcia:branches:security/scap-security-guide

Check the .spec file (which is the RPM packaging) for that package:
https://build.opensuse.org/package/view_file/home:pagarcia:branches:security/scap-security-guide/scap-security-guide.spec?expand=1

And the enabled repositories, since it's a bit special for debbuild-enabled projects: there will be e. g. Debian 10 and Debian 10 debbuild. The "metainformation" for that project (not package!) will show you all the repos in a way you can just copy & paste to your own project:
https://build.opensuse.org/projects/home:pagarcia:branches:security/meta

Another one:
https://build.opensuse.org/package/view_file/systemsmanagement:Uyuni:Master:Debian10-Uyuni-Client-Tools/spacecmd/spacecmd.spec?expand=1
https://build.opensuse.org/projects/systemsmanagement:Uyuni:Master:Debian10-Uyuni-Client-Tools/meta

As a qualification task, I would suggest you branch the `smdba` package and modify the .spec to add debbuild superpowers into it:
https://build.opensuse.org/package/show/systemsmanagement:Uyuni:Master/smdba

Feel free to ask any questions you have!

**Comment by drs-11** [comment]
Thank you for the elaborate response! I will get started with the resources you provided.

I do have some questions:
- Do I need to have OpenSUSE as my developer environment? I currently use Ubuntu for development.
- Will this issue mostly involve using developer tools or will there be some programming as well?
- Should I keep contacting you from this issue for further communications?

**Comment by paususe** [comment]
> Thank you for the elaborate response! I will get started with the resources you provided.
> 
> I do have some questions:
> 
> * Do I need to have OpenSUSE as my developer environment? I currently use Ubuntu for development.

It's not a must but it will certainly make things easier, since you will be able to test things locally, package names will be the same, etc. You can start with openSUSE Leap 15.2 in a VM, no need to replace your main operating system.

> * Will this issue mostly involve using developer tools or will there be some programming as well?

For a fully-functional Uyuni Server on Debian, I am 99% sure some programming will be required, but that's out of scope for this GSoC project:

_It is not required to package or make it run 100% of the features during this GSoC project but at least the core functionality should work._

> * Should I keep contacting you from this issue for further communications?

Sure. Or via the Uyuni mailing lists or chat:
https://www.uyuni-project.org/pages/contact.html

**Comment by drs-11** [comment]
Hey @paususe I [branched](https://build.opensuse.org/package/show/home:drs_11:branches:systemsmanagement:Uyuni:Master/smdba) smdba and modified it a bit to support debbuild. 

From what I've gathered so far is that this issue requires to have all the packages of Uyuni project to have support for debian by modifying the packages for debbuild. This will require to find the unsupported items in the spec file and adding suitable Debian alternatives with the right fields from the Debian policy. I got that right so far?

**Comment by paususe** [comment]
> Hey @paususe I [branched](https://build.opensuse.org/package/show/home:drs_11:branches:systemsmanagement:Uyuni:Master/smdba) smdba and modified it a bit to support debbuild.

Could you please enable the publish flag in https://build.opensuse.org/repositories/home:drs_11:branches:systemsmanagement:Uyuni:Master ?


> From what I've gathered so far is that this issue requires to have all the packages of Uyuni project to have support for debian by modifying the packages for debbuild. This will require to find the unsupported items in the spec file and adding suitable Debian alternatives with the right fields from the Debian policy. I got that right so far?

That's 100% right.

On top of that, you may find some dependency package is available from openSUSE but not on Debian. In that case, my suggestion is to take the openSUSE package and debbuild-ify it. 

While a bit confusing due to the large number of packages introduced as second- and third-level dependencies (which should not be a concern in the case of Debian), this dependency graph might be useful if you want to start checking if any first-level dependencies are missing:
https://github.com/paususe/uyuni-dependencies

**Comment by drs-11** [comment]
> Could you please enable the publish flag in https://build.opensuse.org/repositories/home:drs_11:branches:systemsmanagement:Uyuni:Master ?

Done.

> On top of that, you may find some dependency package is available from openSUSE but not on Debian. In that case, my suggestion is to take the openSUSE package and debbuild-ify it.

So after debbuild-ifying it will need to be accepted by the maintainer of the package?

> While a bit confusing due to the large number of packages introduced as second- and third-level dependencies (which should not be a concern in the case of Debian), this dependency graph might be useful if you want to start checking if any first-level dependencies are missing:

Wow that's a big graph haha!

Okay I think understand most of the things that needs to be done. I think I should probably come up with a plan now and then convert that to a proposal after finalizing the plan with you. That seems the way to go?

**Comment by paususe** [comment]
> > Could you please enable the publish flag in [build.opensuse.org/repositories/home:drs_11:branches:systemsmanagement:Uyuni:Master](https://build.opensuse.org/repositories/home:drs_11:branches:systemsmanagement:Uyuni:Master) ?
> 
> Done.

The package fails to install on Debian 10. You can added the repo and try it by yourself by adding this line to your `/etc/apt/sources.list`

```
https://download.opensuse.org/repositories/home:/drs_11:/branches:/systemsmanagement:/Uyuni:/Master/Debian_debbuild_Debian_10/
```

It seems this section of the .spec is not working. 

```
%{python3_sitelib}/smdba
%{python3_sitelib}/SUSE_Manager_Database_Control-%{version}-py3.6.egg-info
```

Maybe the way python3_sitelib is defined on SUSE systems is not working with debbuild? I suggest to try to run the Python code on a pristine Debian 10 system, minimal installation, and see what happens. Maybe some Python build-requires dependency is missing?

> > On top of that, you may find some dependency package is available from openSUSE but not on Debian. In that case, my suggestion is to take the openSUSE package and debbuild-ify it.
> 
> So after debbuild-ifying it will need to be accepted by the maintainer of the package?

Yes. But initially you should work on your own fork in OBS, so that you are not blocked on waiting for the maintainers of other packages.

> > While a bit confusing due to the large number of packages introduced as second- and third-level dependencies (which should not be a concern in the case of Debian), this dependency graph might be useful if you want to start checking if any first-level dependencies are missing:
> 
> Wow that's a big graph haha!
> 
> Okay I think understand most of the things that needs to be done. I think I should probably come up with a plan now and then convert that to a proposal after finalizing the plan with you. That seems the way to go?

Seems like a plan!

**Comment by vzhestkov** [comment]
@paususe Add me as one of the candidates to be a mentor for this project. I suppose I can show some `debbuild` tricks.
